src/Hosting/Systemd/src/SystemdNotifier.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Net.Sockets;
+
+namespace Microsoft.Extensions.Hosting.Systemd
+{
+    public class SystemdNotifier : ISystemdNotifier
+    {
+        private const string NOTIFY_SOCKET = "NOTIFY_SOCKET";
+
+        private readonly string _socketPath;
+
+        public SystemdNotifier() :
+            this(GetNotifySocketPath())
+        { }
+
+        // For testing
+        internal SystemdNotifier(string socketPath)
+        {
+            _socketPath = socketPath;
+        }
+
+        /// <inheritdoc />
+        public bool IsEnabled => _socketPath != null;
+
+        /// <inheritdoc />
+        public void Notify(ServiceState state)
+        {
+            if (!IsEnabled)
+            {
+                return;
+            }
+
+            using (var socket = new Socket(AddressFamily.Unix, SocketType.Dgram, ProtocolType.Unspecified))
```

Comment:
Is it reasonable to open and close these on demand? Is there a universe where we'd want to keep the connection open because of sending other types of messages?

Reply:
This is only for occasional messaging (now: two messages per application lifetime).
The native sd_notify implementation does the same thing: create and close socket per message.
We can change the scope of the socket to the SystemdNotifier instance if it ever makes sense.

Reply:
Sounds good.